### PR TITLE
Publish test results on Azure pipeline generator

### PR DESCRIPTION
**Pull request by egvimo**
Upates Azure DevOps subgenerator:

- Publish test results for backend and frontend tests
- Publish Cypress screenshots if E2E tests failed

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by DanielFran** [comment]
@egvimo Can you update snapshot files?

File: generators/ci-cd/templates/azure-pipelines.yml.ejs [hunk 1]
```diff
@@ -119,6 +132,12 @@ jobs:
   <%= commented %>    CYPRESS_ENABLE_RECORD: <%= cicdIntegrationsCypressDashboard %>
   <%= commented %>    CYPRESS_PROJECT_ID: $(CYPRESS_PROJECT_ID)
   <%= commented %>    CYPRESS_RECORD_KEY: $(CYPRESS_RECORD_KEY)
+  <%= commented %>- task: PublishPipelineArtifact@1
+  <%= commented %>  inputs:
+  <%= commented %>    artifactName: 'cypress-screenshots'
+  <%= commented %>    targetPath: '$(System.DefaultWorkingDirectory)/build/cypress/screenshots'
```
Comment (mshima) [inline generators/ci-cd/templates/azure-pipelines.yml.ejs:138]:
```suggestion
  <%= commented %>    targetPath: '$(System.DefaultWorkingDirectory)/*/cypress/screenshots'
```
maven uses target

File: generators/ci-cd/templates/azure-pipelines.yml.ejs [hunk 2]
```diff
@@ -82,13 +82,26 @@ jobs:
...
 <%_ if (!skipClient) { _%>
   - task: Npm@1
     inputs:
       command: 'custom'
       customCommand: ' run ci:frontend:test'
     displayName: 'TESTS: frontend'
+  - task: PublishTestResults@2
+    inputs:
+      testResultsFormat: 'JUnit'
+      testResultsFiles: '$(System.DefaultWorkingDirectory)/build/test-results/TESTS-results-jest.xml'
```
Comment (mshima) [inline generators/ci-cd/templates/azure-pipelines.yml.ejs:102]:
```suggestion
      testResultsFiles: '$(System.DefaultWorkingDirectory)/*/test-results/TESTS-results-jest.xml'
```

File: generators/ci-cd/templates/azure-pipelines.yml.ejs (hunk 1 above)
Comment (egvimo) [reply generators/ci-cd/templates/azure-pipelines.yml.ejs:138]: @mshima Thanks for your suggestion. Unfortunately this does not work for all paths (Path does not exist error).

I use Gradle as build system and did not thought about Maven which would have `target`  as output directory.

How about declaring a variable on top of the Pipeline and using it for all paths?

```yaml
variables:
<%_ if (buildToolMaven) { _%>
  buildDirectory: $(Build.SourcesDirectory)/target
<%_ } else if (buildToolGradle) { _%>
  buildDirectory: $(Build.SourcesDirectory)/build
<%_ } _%>
```

Comment (mshima) [reply generators/ci-cd/templates/azure-pipelines.yml.ejs:138]: And now? Using DIST_DIR?

Comment (egvimo) [reply generators/ci-cd/templates/azure-pipelines.yml.ejs:138]: If we use the variables from the pipeline we could then use that variable for all paths:

```yaml
targetPath: '$(outputDirectory)/cypress/screenshots'
```

If we use your approach, then `BUILD_DIR` would be the correct variable.

We could also combine that to:

```yaml
variables:
  outputDirectory: $(Build.SourcesDirectory)/<%= BUILD_DIR %>

# ... and use it ...

targetPath: '$(outputDirectory)/cypress/screenshots'
```

Which one do you prefer?

Comment (mshima) [reply generators/ci-cd/templates/azure-pipelines.yml.ejs:138]: Sorry, my suggestion was wrong.
The var I was looking for is buildDir.
It should have target or build value.

If you think it’s better to extract the variable, ok to me.

Comment (mshima) [reply generators/ci-cd/templates/azure-pipelines.yml.ejs:138]: Your choice.
BUILD_DIR is correct.

**Review by mshima — APPROVED** [review]
LGTM

**Pull request merged**